src/tamm/tensor_impl.hpp

```diff
@@ -1040,14 +1043,29 @@ class DenseTensorImpl: public TensorImpl<T> {
   }
 
   void add(const IndexVector& blockid, span<T> buff_span) {
-#if defined(USE_UPCXX)
-    throw std::runtime_error("upcxx: dense tensor - add unsupported");
-#else
     EXPECTS(allocation_status_ != AllocationStatus::invalid);
-    std::vector<int64_t> lo = compute_lo(blockid);
-    std::vector<int64_t> hi = compute_hi(blockid);
-    std::vector<int64_t> ld = compute_ld(blockid);
     EXPECTS(block_size(blockid) <= buff_span.size());
+
+    std::vector<int64_t> lo = compute_lo(blockid);
+
+#if defined(USE_UPCXX)
+    for(size_t i = lo.size(); i < 4; ++i) lo.insert(lo.begin(), 0);
+
+    TensorTile t = find_tile(lo[0], lo[1], lo[2], lo[3]);
+
+    upcxx::rpc(
+      *ec_->pg().team(), t.rank,
+      [](const upcxx::global_ptr<T>& dst_buf, const upcxx::view<T>& src_buf) {
+        T*     dst = dst_buf.local();
+        size_t n   = src_buf.size();
+
+        for(size_t i = 0; i < n; ++i) dst[i] += src_buf[i];
+      },
+      upcxx::reinterpret_pointer_cast<T>(gptrs_[t.rank]) + t.offset,
+      upcxx::make_view((T*) buff_span.data(), (T*) buff_span.data() + block_size(blockid)))
+      .wait();
```

Comment:
This is a fully blocking implementation of accumulate, ie this `wait()` call stalls here for the full round-trip latency of the RPC.

Based on my understanding of [GA semantics](https://hpc.pnl.gov/globalarrays/papers/GA-UserManual-Main.pdf), the [`NGA_Acc64()`](https://hpc.pnl.gov/globalarrays/api/c_op_api.html#ACC) call below in the GA version is non-blocking and remains in-flight asynchronously until a subsequent `GA_Fence()` or `GA_Sync()` call from this rank. 

I'd really like to see us move towards more asynchronous communication (something that UPC++ supports _very_ well), to improve performance via overlap. I don't know the extent to which the latency of this routine matters for end-to-end performance, but I suspect it matters.

@jcorbino Do you have a plan to go back and remove all the `future::wait()` calls from code paths that matter to performance, or should we be designing that in starting now?

Reply:
@bonachea Thanks for your quick feedback. Here's what I think:

#1 `NGA_Acc64()` is a blocking call
#2 This operation (add) should not be by default nonblocking, the calling processes expect the addition to be done before proceeding
#3 The next phase (once I'm done with completeness) is to improve performance

**Of course, we should exploit asynchrony as much as possible**, but in this particular fragment, we're just adding a buffer in a blocking way.

Reply:
`NGA_Acc64` definitely locks for _local_ completion (safe to overwrite source memory), but not remote completion (result committed to remote memory).  What you've written for UPC++ blocks for _remote_ completion, which takes much longer, and I suspect is unnecessary to fulfill the semantics of this function, thus leaving performance on the table.

Sorry I was unclear about that in my initial comment.

Reply:
Thank you @bonachea , as we discussed via Slack, I will update this (and other) routines where asynchrony can be exploited. I will make a separate PR just for that. For now, this is good enough for completeness and consistency with respect to other parts of the code.